### ISSUE TITLE
Discussion: Optimization for large AppendEntries RPC

### ISSUE THREAD

**Issue by xu-cheng** [async-raft/async-raft#103]
This is a more of feature request on a possible optimization.

Currently, the AppendEntries RPC has a timeout guard using `heartbeat_interval` from the `Config`. In my use case, the latency between nodes is small, so ideally we should use a small `heartbeat_interval` for better performance. At the same time, the size of `AppData` and `AppDataResponse` can be quite large. This entails a very large `heartbeat_interval`. Otherwise, the system will report "timeout while sending AppendEntries RPC to target" and cause view change and leader election.  I think this results to a suboptimal performance.

May I suggest to introduce a new `append_entries_timeout` config? In this case, when sending AppendEntries RPC, we should use `append_entries_timeout` as the timeout for normal AppendEntries RPC and `heartbeat_interval` as the timeout for heartbeat RPC.

**Comment by xu-cheng** [comment]
More generally, would it be a good idea to separate the AppendEntries RPC and the heartbeat RPC?

**Comment by thedodd** [comment]
> More generally, would it be a good idea to separate the AppendEntries RPC and the heartbeat RPC?

@xu-cheng great question. I don't think we should have two separate RPCs, as that would diverge from the spec. Also, this is a fairly central configuration issue in Raft overall. Here is a link to the docs in our code base which is just quoting the Raft spec: https://github.com/async-raft/async-raft/blob/master/async-raft/src/config.rs#L49-L65

```
/// When building the Raft configuration for your application, remember this inequality from the
/// Raft spec: `broadcastTime ≪ electionTimeout ≪ MTBF`.
///
/// > In this inequality `broadcastTime` is the average time it takes a server to send RPCs in
/// > parallel to every server in the cluster and receive their responses; `electionTimeout` is the
/// > election timeout described in Section 5.2; and `MTBF` is the average time between failures for
/// > a single server. The broadcast time should be an order of magnitude less than the election
/// > timeout so that leaders can reliably send the heartbeat messages required to keep followers
/// > from starting elections; given the randomized approach used for election timeouts, this
/// > inequality also makes split votes unlikely. The election timeout should be a few orders of
/// > magnitude less than `MTBF` so that the system makes steady progress. When the leader crashes,
/// > the system will be unavailable for roughly the election timeout; we would like this to
/// > represent only a small fraction of overall time.
///
/// What does all of this mean? Simply keep your election timeout settings high enough that the
/// performance of your network will not cause election timeouts, but don't keep it so high that
/// a real leader crash would cause prolonged downtime. See the Raft spec §5.6 for more details.
```

@xu-cheng when you say "so ideally we should use a small `heartbeat_interval` for better performance", what exactly do you mean? What performance are you referring to? Replication is not limited by the heartbeat interval. The system will replicate as fast as possible, send then next payload of entries as soon as the previously sent payload has been processed by the follower.

**Comment by xu-cheng** [comment]
> when you say "so ideally we should use a small heartbeat_interval for better performance", what exactly do you mean? What performance are you referring to? 

I mean the performance in terms of the election time. Currently, in my case, the request is quite large so the heartbeat interval need to be around 10~30 seconds. However, this results to election time lasting at least one minute. I am wondering if it is possible to reduce such a delay for both the network startup or when a view change happens.

**Comment by thedodd** [comment]
Very interesting. So, the only thing which blocks the immediate response is the process of writing the logs to the storage engine: https://github.com/async-raft/async-raft/blob/master/async-raft/src/core/append_entries.rs#L157 So in this implementation of Raft the only overhead the replication process has is network transmission & speed of the storage engine to write a log. This does not involve writing the log entry to the state machine, only appending to the log.

10~30s seems like a really long time to transmit data. It sounds to me like that is a fairly serious issue. If it is taking that long, then either your storage engine is SUPER slow or you are transferring log entires which are literally multiple GBs in size.

In any of those cases, there are things which can be done to keep the replication process and log entry size smaller. I would investigate a bit on that front. Do you mind me asking the average size of these large entries? What type of data is being transmitted?

**Comment by xu-cheng** [comment]
I am using async-raft to build a blockchain prototype for my research project. I try to find ways to achieve highest possible throughput. (I probably should use a PBFT consensus library, but couldn’t find a good rust implementation). 

The average size of the log entries is about 50~100MB. In my setup, there are 16 nodes, so the total amount of network transmission per log entry could be in GB (100MB x 16). 

> This does not involve writing the log entry to the state machine, only appending to the log.

Isn’t the next line of code after https://github.com/async-raft/async-raft/blob/master/async-raft/src/core/append_entries.rs#L157 related to state machine. 

In my settings, there is further slow down 
when writing to state machine. It takes a few seconds.

**Comment by xu-cheng** [comment]
Yes, I can reduce the size of the log entries. But it would result to lower throughputs.

**Comment by thedodd** [comment]
> Isn’t the next line of code after https://github.com/async-raft/async-raft/blob/master/async-raft/src/core/append_entries.rs#L157 related to state machine.

Yes, however if you look at that function, it does not block the task, it simply updates the entry cache, updates a few pointers & then spawns a separate task which handles the actual replication to the state machine. So the AppendEntries RPC handler itself is not blocked by that process. The only thing which blocks the AppendEntries RPC task is the appending of the entries to the log, which is fundamentally required by Raft.

> I probably should use a PBFT consensus library, but couldn’t find a good rust implementation

For blockchain stuff yea, but I'm glad you are using this project! You've been providing a lot of great feedback and making lots of great updates to the system, so thank you!

> The average size of the log entries is about 50~100MB. In my setup, there are 16 nodes, so the total amount of network transmission per log entry could be in GB (100MB x 16).

This does seem to indicate that we will want to have a separate mechanism for timeouts on heartbeats vs AppendEntries RPCs which have payloads.
- if you take a look at the code here (which we recently updated) https://github.com/async-raft/async-raft/blob/master/async-raft/src/core/append_entries.rs#L16-L59 you will see that heartbeats are not evaluated deeply, so we could introduce an independent loop in the replication streams which will always send heartbeats outside of the context of actual entry replication.
- if we did that, then we would be able to introduce a separate timeout for heartbeats vs entries. This would preserve quick detection of a dead leader by follower nodes and would also allow for replication of very large entries. 

Thoughts?

**Comment by xu-cheng** [comment]
I mainly want to raise this issue for discussion, as it comes from a research project not real world applications. Therefore, it is perfectly fine to me if this issue is unsolved. 

I also tried to search research papers on Raft to see whether there is an established improvement/extension to the Raft protocol to address this issue. So far I didn’t find one.  

It is probably a better idea to not divergent the implementation from the specs. Especially considering that one of the advantages of the Raft and Paxos is that they have passed rigorous formal analysis. 

This is just my two cents.

**Comment by thedodd** [comment]
Agreed, however even if we make the changes I've mentioned above, our algorithm would still fully comply with the Raft spec. 

However, there is an easier path forward.
- Don't embed the large data payloads in the Raft RPCs, only embed pointers.
- Each node should have a network endpoint which can receive a request to download a chunk by UUID (or the like).
- The log replication process will just embed the pointers to these chunks, and the follower's handler to append entries to the log (the RaftStorage impl specifically, which you control in your code) can just start the process of downloading these chunks from the leader.
- The leader will see most of the requests as timing out, but the process of downloading can be asynchronous and not bound to the lifetime of the RPC. Once the chunk has been downloaded from the leader, the next RPC instructing for the chunk to be downloaded will succeed, and that entry can then be safely appended to the follower's log.

Fully Raft compliant, log is safe as log appending does not actually succeed until the chunk is downloaded. Snapshotting will have to be updated to work with this pattern as well, but that should be fine. 

Thoughts?

**Comment by xu-cheng** [comment]
Thanks for the advice. It is a good idea to move large message transmission outside Raft.
